src/Microsoft.AspNet.Mvc.Core/Formatters/EnumerableWrapperProvider.cs

```diff
@@ -0,0 +1,135 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Reflection;
+
+namespace Microsoft.AspNet.Mvc
+{
+    public class EnumerableWrapperProvider : IWrapperProvider
+    {
+        private static readonly ConcurrentDictionary<Type, Type> _delegatingEnumerableCache = new ConcurrentDictionary<Type, Type>();
+        private static ConcurrentDictionary<Type, ConstructorInfo> _delegatingEnumerableConstructorCache = new ConcurrentDictionary<Type, ConstructorInfo>();
+        private readonly IEnumerable<IWrapperProvider> _wrapperProviders;
+
+        public EnumerableWrapperProvider(IEnumerable<IWrapperProvider> wrapperProviders)
+        {
+            _wrapperProviders = wrapperProviders;
+        }
+
+        public bool TryGetWrappingTypeForDeserialization(Type originalType, out Type wrappingType)
+        {
+            wrappingType = null;
+
+            return false;
+        }
+
+        // Examples: 
+        // a. IEnumerable<SerializableError> => DelegatingEnumerable<SerializableErrorWrapper>
+        // b. IEnumerable<Person> => DelegatingEnumerable<Person>
+        public bool TryGetWrappingTypeForSerialization(Type originalType, out Type wrappingType)
+        {
+            return TryGetDelegatingType(originalType, out wrappingType);
+        }
+
+        /// <inheritdoc />
+        public object Unwrap(Type declaredType, object obj)
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <inheritdoc />
+        public object Wrap(Type declaredType, object original)
+        {
+            Type delegatingType;
+            if (TryGetDelegatingType(declaredType, out delegatingType))
```

Comment:
I see we are caching why the delegating type is but not any of the constructor infos/Invocation etc. Seems like we are caching at the wrong layer.

This in general requires at the least a perf analysis with a profiler before proceeding